### ISSUE TITLE
Issues with installation

### ISSUE THREAD

**Issue by pwh124** [dieterich-lab/single-cell-nanopore#9]
Hello,

I am trying to install scnapbar software with guide in the README.

However I am encountering the following error when running make:

> [ 50%] Building CXX object src/CMakeFiles/singleCellPipe.dir/Flexbar.cpp.o
In file included from /home/alicecsq/software/scnapbar/src/Flexbar.cpp:24:
/home/alicecsq/software/scnapbar/src/Flexbar.h:15:10: fatal error: tbb/pipeline.h: No such file or directory
  15 | #include <tbb/pipeline.h>
   |     ^~~~~~~~~~~~~~~~
compilation terminated.

Do you have any suggestions as to how we can fix this? Does Flexbar need to be installed manually? Do we need a specific version of python?

Thanks!
Paul

>

**Comment by qwang-big** [comment]
Hi Paul,

You need to install the packages libtbb-dev and libtbb2 in order to compile. 
The requirement is the same as Flexbar, see their description:
https://github.com/seqan/flexbar/blob/master/README.md#building-from-source

Python is used in the Snakemake pipeline, you may cCreate the environment from the environment.yaml file:

conda env create -f environment.yaml

Qi

**Comment by pwh124** [comment]
Hello - thanks for responding... I have followed the download instructions from the README so both tbb and tbb-devel are installed. 

I get the same error that it cannot find tbb/pipeline.h even though through downloading tbb, I do have that script.

Paul

**Comment by qwang-big** [comment]
Hi Paul,

I can reproduce your error on a freshly installed ubuntu system. The error can be solved by the following commands in the `conda` enviroment
```
conda install tbb tbb-devel
```
please give a try

Qi

**Comment by pwh124** [comment]
Hi Qi,

Thanks for helping... I think I figured it out. The problem wasn't TBB... the problem was the compiler (I think). Warning: long answer ahead.

### TL;DR: TBB was not the issue, the compiler was. Installing `gxx_linux-64` fixed the issue. However another issue with seqan arose

#### First I created a clean conda environment with python 3.7.6 (the same in your environment). I then activated it, installed TBB, and then downloaded the github repo.
```bash
conda create --name single-cell-nanopore python=3.7.6
conda activate single-cell-nanopore
conda install tbb tbb-devel
git clone https://github.com/dieterich-lab/single-cell-nanopore.git
```

#### I then moved into the Github repo and ran cmake
```bash
cd single-cell-nanopore/
cmake .
```
#### Which gave this output
> -- The C compiler identification is GNU 7.5.0
-- The CXX compiler identification is GNU 7.5.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test COMPILER_SUPPORTS_CXX14
-- Performing Test COMPILER_SUPPORTS_CXX14 - Success
-- Flexbar 64 bit architecture
-- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version "1.2.11") 
-- Found BZip2: /usr/lib/x86_64-linux-gnu/libbz2.so (found version "1.0.6") 
-- Looking for BZ2_bzCompressInit
-- Looking for BZ2_bzCompressInit - found
-- Configuring done
-- Generating done
-- Build files have been written to: /home/phook2/single-cell-nanopore

#### I then ran make
```bash
make
```
#### And got this output (which is the error I have been seeing)
> Scanning dependencies of target singleCellPipe
[ 50%] Building CXX object src/CMakeFiles/singleCellPipe.dir/Flexbar.cpp.o
In file included from /home/phook2/single-cell-nanopore/src/Flexbar.cpp:24:0:
/home/phook2/single-cell-nanopore/src/Flexbar.h:15:10: fatal error: tbb/pipeline.h: No such file or directory
 #include <tbb/pipeline.h>
          ^~~~~~~~~~~~~~~~
compilation terminated.
src/CMakeFiles/singleCellPipe.dir/build.make:62: recipe for target 'src/CMakeFiles/singleCellPipe.dir/Flexbar.cpp.o' failed
make[2]: *** [src/CMakeFiles/singleCellPipe.dir/Flexbar.cpp.o] Error 1
CMakeFiles/Makefile2:85: recipe for target 'src/CMakeFiles/singleCellPipe.dir/all' failed
make[1]: *** [src/CMakeFiles/singleCellPipe.dir/all] Error 2
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2

#### When I was playing around with trying to install this in my base conda environment, I noticed that when I tried to re-install TBB this line after `cmake` changed from this: ```-- Check for working C compiler: /usr/bin/cc -- works``` to this `-- Check for working C compiler: /home/phook2/miniconda3/envs/single-cell-nanopore/bin/x86_64-conda_cos6-linux-gnu-cc -- works` - so I installed ```gxx_linux-64```

```bash
conda install gxx_linux-64
```
#### I then re-downloaded a clean single-cell-nanopore repo
```bash
cd ..
rm -r single-cell-nanopore/
git clone https://github.com/dieterich-lab/single-cell-nanopore.git
```
#### Moved into dir and ran cmake  again
```bash
#### I then moved into the Github repo and ran cmake
```bash
cd single-cell-nanopore/
cmake .
```
#### And got this output 
> -- The C compiler identification is GNU 7.3.0
-- The CXX compiler identification is GNU 7.3.0
-- Check for working C compiler: /home/phook2/miniconda3/envs/single-cell-nanopore/bin/x86_64-conda_cos6-linux-gnu-cc
-- Check for working C compiler: /home/phook2/miniconda3/envs/single-cell-nanopore/bin/x86_64-conda_cos6-linux-gnu-cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /home/phook2/miniconda3/envs/single-cell-nanopore/bin/x86_64-conda_cos6-linux-gnu-c++
-- Check for working CXX compiler: /home/phook2/miniconda3/envs/single-cell-nanopore/bin/x86_64-conda_cos6-linux-gnu-c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test COMPILER_SUPPORTS_CXX14
-- Performing Test COMPILER_SUPPORTS_CXX14 - Success
-- Flexbar 64 bit architecture
-- Found ZLIB: /home/phook2/miniconda3/envs/single-cell-nanopore/lib/libz.so (found version "1.2.11") 
-- Could NOT find BZip2 (missing: BZIP2_LIBRARIES) (found version "1.0.6")
-- Build will not support bzip2.
-- Configuring done
-- Generating done
-- Build files have been written to: /home/phook2/single-cell-nanopore

#### I then ran make
```bash
make
```
#### And got this output:
> Scanning dependencies of target singleCellPipe
[ 50%] Building CXX object src/CMakeFiles/singleCellPipe.dir/Flexbar.cpp.o
In file included from /home/phook2/single-cell-nanopore/src/Flexbar.cpp:24:0:
/home/phook2/single-cell-nanopore/src/Flexbar.h:20:10: fatal error: seqan/basic.h: No such file or directory
 #include <seqan/basic.h>
          ^~~~~~~~~~~~~~~
compilation terminated.
src/CMakeFiles/singleCellPipe.dir/build.make:62: recipe for target 'src/CMakeFiles/singleCellPipe.dir/Flexbar.cpp.o' failed
make[2]: *** [src/CMakeFiles/singleCellPipe.dir/Flexbar.cpp.o] Error 1
CMakeFiles/Makefile2:85: recipe for target 'src/CMakeFiles/singleCellPipe.dir/all' failed
make[1]: *** [src/CMakeFiles/singleCellPipe.dir/all] Error 2
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2

#### Not a TBB error any more but a seqan error. I think this can be fixed by manually installing seqan and moving "include/" into the "seqan" dir in the "single-cell-nanopore" dir

**Comment by pwh124** [comment]
I've solved the issues and I am now able to build the software.

```bash
# Download scnapbar repo
git clone https://github.com/dieterich-lab/single-cell-nanopore.git

# Download seqan and move into `seqan` dir in scnapbar repo. This was tried because of suggestions from flexbar documentation: https://github.com/seqan/flexbar
wget https://github.com/seqan/seqan/releases/download/seqan-v2.4.0/seqan-library-2.4.0.tar.xz
tar xJf seqan-library-2.4.0.tar.xz
cp -r seqan-library-2.4.0/include single-cell-nanopore/seqan/

# Modify the environment.yaml file
cd single-cell-nanopore/
cp environment.yaml good.environment.yaml
```
Change/add the following to `environment.yaml`:
> - mummer4=4.0.0beta2 # changed from mummer=4.0.0beta2
> - r-markdown=1.1 # changed from r::r-markdown=1.1
> - tbb=2020.3 # Added. Note, compiling would not work with 2021.1 version downloaded from conda-forge
> - tbb-devel=2020.3 # Added. Note, compiling would not work with 2021.1 version downloaded from conda-forge

Create conda environment with new environment. Conda version 4.9.2
```bash
conda env create --name single-cell-nanopore --file good.environment.yaml
conda activate single-cell-nanopore
```
Build the software
```bash
cmake .
make
```
The software should successfully build with some messages about deprecated TBB commands.

**Comment by qwang-big** [comment]
thanks Paul, these are very helpful to the other users.

**Comment by eboileau** [comment]
For the record:

one can add  the `--recursive` flag when cloning the repo to include the seqan submodule.
If the repo was cloned without the flag, one can run afterwards -- but before building/running make --

```
git submodule update --init 
```

to fetch the seqan submodule.



As a follow-up on the previous remark about _deprecated TBB commands_, there are indeed the following warnings (trimmed output):

```
In file included from /prj/Bartenschlager/analysis/single-cell-nanopore/src/Flexbar.h:16,
                 from /prj/Bartenschlager/analysis/single-cell-nanopore/src/Flexbar.cpp:24:
/home/eboileau/.miniconda3/envs/scNapBar/include/tbb/task_scheduler_init.h:21:154: note: #pragma message: TBB Warning: tbb/task_scheduler_init.h is deprecated. 

In file included from /prj/Bartenschlager/analysis/single-cell-nanopore/src/Flexbar.h:18,
                 from /prj/Bartenschlager/analysis/single-cell-nanopore/src/Flexbar.cpp:24:
/home/eboileau/.miniconda3/envs/scNapBar/include/tbb/mutex.h:21:140: note: #pragma message: TBB Warning: tbb/mutex.h is deprecated. 

In file included from /prj/Bartenschlager/analysis/single-cell-nanopore/src/Flexbar.cpp:24:
/prj/Bartenschlager/analysis/single-cell-nanopore/src/Flexbar.h: In function 'void startProcessing(Options&, AlignmentResults&)':
/prj/Bartenschlager/analysis/single-cell-nanopore/src/Flexbar.h:247:16: warning: 'pipeline' is deprecated: tbb::pipeline is deprecated, use tbb::parallel_pipeline [-Wdeprecated-declarations] 247 |  tbb::pipeline pipe;
                                                                        ^~~~
In file included from /prj/Bartenschlager/analysis/single-cell-nanopore/src/Flexbar.h:15,
                 from /prj/Bartenschlager/analysis/single-cell-nanopore/src/Flexbar.cpp:24:
/home/eboileau/.miniconda3/envs/scNapBar/include/tbb/pipeline.h:235:87: note: declared here
235 | class __TBB_DEPRECATED_MSG("tbb::pipeline is deprecated, use tbb::parallel_pipeline") pipeline {
                                                                                            ^~~~
In file included from /prj/Bartenschlager/analysis/single-cell-nanopore/src/Flexbar.cpp:26:
/prj/Bartenschlager/analysis/single-cell-nanopore/src/removeCDNA.h: At global scope:
/prj/Bartenschlager/analysis/single-cell-nanopore/src/removeCDNA.h:86:30: warning: use of 'auto' in parameter declaration only available with '-fconcepts'
86 | Dna5String & getOriginalRead(auto & readMap, string & readID)
                                  ^~~~
```
